File: source/Game/World.lua
class('World').extends()

function World:init(size)
    self.size = size or 200
    self.particles = {}
    
    -- Cube Vertices
    self.vertices = {
        Vector3(-size, -size, -size), Vector3(size, -size, -size),
        Vector3(size, size, -size), Vector3(-size, size, -size),
        Vector3(-size, -size, size), Vector3(size, -size, size),
        Vector3(size, size, size), Vector3(-size, size, size)
    }
    
    -- Cube Edges (indices)
    self.edges = {
        {1,2}, {2,3}, {3,4}, {4,1}, -- Back face (Z-)
        {5,6}, {6,7}, {7,8}, {8,5}, -- Front face (Z+)
        {1,5}, {2,6}, {3,7}, {4,8}  -- Connecting edges
    }
    
    -- Water Surface Grid (Top Face: Y = +size)
    self.surfaceRows = 4
    self.surfaceCols = 4
    self.surfacePoints = {}
    local stepX = (size * 2) / self.surfaceRows
    local stepZ = (size * 2) / self.surfaceCols
    
    for x = 0, self.surfaceRows do
        for z = 0, self.surfaceCols do
            local px = -size + x * stepX
            local pz = -size + z * stepZ
            table.insert(self.surfacePoints, {base = Vector3(px, size, pz), current = Vector3(px, size, pz)})
        end
    end
    
    -- -- Caustic Particles (Bottom Face: Y = -size)
    -- self.caustics = {}
    -- for i=1, 40 do
    --     table.insert(self.caustics, {
    --         pos = Vector3(math.random(-size, size), -size, math.random(-size, size)),
    --         speed = math.random() * 0.05 + 0.02,
    --         offset = math.random() * math.pi * 2
    --     })
    -- end
    
    -- Suspended Particles
    math.randomseed(playdate.getSecondsSinceEpoch())
    for i=1, 60 do
        table.insert(self.particles, Vector3(
            math.random(-size, size), -- X
            math.random(-size, size), -- Y
            math.random(-size, size)  -- Z
        ))
    end
    
    self.accumulatedWaveDt = 0
end

function World:draw(camera, dt)
    local gfx = playdate.graphics
    local time = playdate.getCurrentTimeMilliseconds() / 1000.0
    
    -- Update Surface Waves & Refraction
    self.accumulatedWaveDt = self.accumulatedWaveDt + dt
    if self.accumulatedWaveDt >= 1.0 then
        for _, p in ipairs(self.surfacePoints) do
            -- Vertical Wave (Height)
            local waveY = math.sin(p.base.x * 0.03 + time * 2.5) * 8 
                        + math.cos(p.base.z * 0.04 + time * 2.0) * 8
            
            -- Horizontal Refraction (Fake light bending)
            -- We distort X based on Z, and Z based on X to create swirling feel
            local refractX = math.sin(p.base.z * 0.05 + time * 3.0) * 12
            local refractZ = math.cos(p.base.x * 0.05 + time * 2.5) * 12
            
            p.current.y = p.base.y + waveY
            p.current.x = p.base.x + refractX
            p.current.z = p.base.z + refractZ
        end
        self.accumulatedWaveDt = 0
    end
    
    -- Project Vertices
    local projected = {}
    for i, v in ipairs(self.vertices) do
        projected[i] = camera:project(v)
    end
    
    -- Draw Cube Edges
    gfx.setColor(gfx.kColorBlack)
    gfx.setLineWidth(2)
    
    for _, edge in ipairs(self.edges) do
        local p1 = projected[edge[1]]
        local p2 = projected[edge[2]]
        
        if p1 and p2 then
            gfx.drawLine(p1.x, p1.y, p2.x, p2.y)
        end
    end
    
    -- Draw Surface Grid (Waves) with Dithering
    gfx.setLineWidth(1)
    local idx = 1
    local cols = self.surfaceCols + 1
    for x = 0, self.surfaceRows do
        for z = 0, self.surfaceCols do
            local p = self.surfacePoints[idx].current
            local proj = camera:project(p)
            
            if proj then 
                -- Calculate wave intensity for dithering
                local waveIntensity = math.abs(p.y - self.surfacePoints[idx].base.y) / 16
                local distortion = math.abs(p.x - self.surfacePoints[idx].base.x) / 12
                local totalDistortion = waveIntensity + distortion
                
                -- Apply dithering based on distortion level
                if totalDistortion > 1.5 then
                    gfx.setDitherPattern(0.7, gfx.image.kDitherTypeBayer4x4)
                elseif totalDistortion > 1.0 then
                    gfx.setDitherPattern(0.5, gfx.image.kDitherTypeBayer4x4)
                elseif totalDistortion > 0.5 then
                    gfx.setDitherPattern(0.3, gfx.image.kDitherTypeBayer8x8)
                else
                    gfx.setColor(gfx.kColorBlack)
                end
                
                -- Draw lines to neighbors
                if z < self.surfaceCols then -- Connect Z neighbor
                    local nextZ = self.surfacePoints[idx + 1].current
                    local nextProj = camera:project(nextZ)
                    if nextProj then gfx.drawLine(proj.x, proj.y, nextProj.x, nextProj.y) end
                end
                
                if x < self.surfaceRows then -- Connect X neighbor
                    local nextX = self.surfacePoints[idx + cols].current
                    local nextProj = camera:project(nextX)
                    if nextProj then gfx.drawLine(proj.x, proj.y, nextProj.x, nextProj.y) end
                end
            end
            idx = idx + 1
        end
    end
    
    -- Reset color for next drawing
    gfx.setColor(gfx.kColorBlack)
    
    -- -- Draw Floor Caustics (simulated light patterns)
    -- for _, c in ipairs(self.caustics) do
    --     -- Animate caustic
    --     c.pos.x = c.pos.x + math.sin(time * 2.0 + c.offset) * 1.0
    --     c.pos.z = c.pos.z + math.cos(time * 1.5 + c.offset) * 1.0
        
    --     -- Wrap around
    --     if c.pos.x > self.size then c.pos.x = -self.size end
    --     if c.pos.x < -self.size then c.pos.x = self.size end
    --     if c.pos.z > self.size then c.pos.z = -self.size end
    --     if c.pos.z < -self.size then c.pos.z = self.size end
        
    --     local proj = camera:project(c.pos)
    --     if proj then
    --         -- Caustics are bright, so maybe use white with dithering or just circles?
    --         -- Since background is white (screen clear), we draw black.
    --         -- To make it look like light, it should be inverted? 
    --         -- Playdate is 1-bit. Light = White, Shadow = Black.
    --         -- So "Caustics" (Light) should be White on dark? Or Black patterns on White?
    --         -- Standard Playdate: White background.
    --         -- "Caustics" are usually bright lines.
    --         -- If we draw black circles, they look like shadows.
    --         -- Let's draw rapidly changing dithered circles to simulate shimmering light patterns?
    --         -- Or simple rings.
            
    --         local r = math.max(2, 100 / proj.z)
    --         if r > 15 then r = 15 end
            
    --         -- Draw a "light ring"
    --         gfx.setLineWidth(2)
    --         gfx.drawCircleAtPoint(proj.x, proj.y, r)
            
    --         -- Wobbly inner
    --         -- gfx.fillCircleAtPoint(proj.x, proj.y, r/2)
    --     end
    -- end
    
    -- Draw Particles
    gfx.setLineWidth(1)
    for _, p in ipairs(self.particles) do
        local proj = camera:project(p)
        if proj then
            -- Depth Dithering
            -- Closer particles are darker/solid
            -- Far particles are lighter/dithered
            
            local z = proj.z
            local radius = math.max(1, 200 / z)
            if radius > 8 then radius = 8 end
            
            if z < 100 then
                 gfx.setColor(gfx.kColorBlack)
            elseif z < 200 then
                gfx.setDitherPattern(0.5, gfx.image.kDitherTypeBayer4x4)
            elseif z < 300 then
                gfx.setDitherPattern(0.2, gfx.image.kDitherTypeBayer8x8)
            else
                gfx.setDitherPattern(0.1, gfx.image.kDitherTypeBayer8x8)
            end
            
            gfx.fillCircleAtPoint(proj.x, proj.y, radius)
        end
    end
    
    -- Draw Boundary Warning
    if math.abs(camera.position.x) >= self.size - 5 or
       math.abs(camera.position.y) >= self.size - 5 or
       math.abs(camera.position.z) >= self.size - 5 then
       
       gfx.setColor(gfx.kColorBlack)
       gfx.drawText("WALL HIT", 5, 220)
    end
end

File: source/Game/World.c
#include "World.h"

#include <math.h>
#include <stdlib.h>

static float random_range(float min_value, float max_value) {
    const float random_unit = (float)rand() / (float)RAND_MAX;
    return min_value + random_unit * (max_value - min_value);
}

static void world_update_surface(World* world, float elapsed_seconds, float dt) {
    world->accumulated_wave_dt += dt;

    if (world->accumulated_wave_dt < 1.0f) {
        return;
    }

    for (int i = 0; i < WORLD_SURFACE_POINT_COUNT; ++i) {
        SurfacePoint* point = &world->surface_points[i];

        const float wave_y = sinf(point->base.x * 0.03f + elapsed_seconds * 2.5f) * 8.0f
                           + cosf(point->base.z * 0.04f + elapsed_seconds * 2.0f) * 8.0f;

        const float refract_x = sinf(point->base.z * 0.05f + elapsed_seconds * 3.0f) * 12.0f;
        const float refract_z = cosf(point->base.x * 0.05f + elapsed_seconds * 2.5f) * 12.0f;

        point->current.x = point->base.x + refract_x;
        point->current.y = point->base.y + wave_y;
        point->current.z = point->base.z + refract_z;
    }

    world->accumulated_wave_dt = 0.0f;
}

void world_init(World* world, int size) {
    world->size = size;
    world->accumulated_wave_dt = 0.0f;

    const float s = (float)size;

    world->vertices[0] = vector3_create(-s, -s, -s);
    world->vertices[1] = vector3_create(s, -s, -s);
    world->vertices[2] = vector3_create(s, s, -s);
    world->vertices[3] = vector3_create(-s, s, -s);
    world->vertices[4] = vector3_create(-s, -s, s);
    world->vertices[5] = vector3_create(s, -s, s);
    world->vertices[6] = vector3_create(s, s, s);
    world->vertices[7] = vector3_create(-s, s, s);

    const int edges[WORLD_EDGE_COUNT][2] = {
        {0, 1}, {1, 2}, {2, 3}, {3, 0},
        {4, 5}, {5, 6}, {6, 7}, {7, 4},
        {0, 4}, {1, 5}, {2, 6}, {3, 7}
    };

    for (int i = 0; i < WORLD_EDGE_COUNT; ++i) {
        world->edges[i][0] = edges[i][0];
        world->edges[i][1] = edges[i][1];
    }

    const float step_x = (s * 2.0f) / (float)WORLD_SURFACE_ROWS;
    const float step_z = (s * 2.0f) / (float)WORLD_SURFACE_COLS;
    int index = 0;

    for (int x = 0; x <= WORLD_SURFACE_ROWS; ++x) {
        for (int z = 0; z <= WORLD_SURFACE_COLS; ++z) {
            const float point_x = -s + (float)x * step_x;
            const float point_z = -s + (float)z * step_z;
            const Vector3 point = vector3_create(point_x, s, point_z);
            world->surface_points[index].base = point;
            world->surface_points[index].current = point;
            ++index;
        }
    }

    for (int i = 0; i < WORLD_PARTICLE_COUNT; ++i) {
        world->particles[i] = vector3_create(
            random_range(-s, s),
            random_range(-s, s),
            random_range(-s, s)
        );
    }
}

void world_draw(World* world, const Camera* camera, PlaydateAPI* pd, float dt) {
    const float elapsed_seconds = (float)pd->system->getCurrentTimeMilliseconds() / 1000.0f;
    world_update_surface(world, elapsed_seconds, dt);

    pd->graphics->setLineCapStyle(kLineCapStyleRound);

    ProjectedPoint projected_vertices[WORLD_VERTEX_COUNT];
    int is_visible[WORLD_VERTEX_COUNT];

    for (int i = 0; i < WORLD_VERTEX_COUNT; ++i) {
        is_visible[i] = camera_project(camera, world->vertices[i], &projected_vertices[i]);
    }

    for (int i = 0; i < WORLD_EDGE_COUNT; ++i) {
        const int a = world->edges[i][0];
        const int b = world->edges[i][1];

        if (is_visible[a] && is_visible[b]) {
            pd->graphics->drawLine(
                (int)projected_vertices[a].x,
                (int)projected_vertices[a].y,
                (int)projected_vertices[b].x,
                (int)projected_vertices[b].y,
                2,
                kColorBlack
            );
        }
    }

    int idx = 0;
    const int cols = WORLD_SURFACE_COLS + 1;

    for (int x = 0; x <= WORLD_SURFACE_ROWS; ++x) {
        for (int z = 0; z <= WORLD_SURFACE_COLS; ++z) {
            ProjectedPoint current_projection;
            if (camera_project(camera, world->surface_points[idx].current, &current_projection)) {
                if (z < WORLD_SURFACE_COLS) {
                    ProjectedPoint next_projection_z;
                    if (camera_project(camera, world->surface_points[idx + 1].current, &next_projection_z)) {
                        pd->graphics->drawLine(
                            (int)current_projection.x,
                            (int)current_projection.y,
                            (int)next_projection_z.x,
                            (int)next_projection_z.y,
                            1,
                            kColorBlack
                        );
                    }
                }

                if (x < WORLD_SURFACE_ROWS) {
                    ProjectedPoint next_projection_x;
                    if (camera_project(camera, world->surface_points[idx + cols].current, &next_projection_x)) {
                        pd->graphics->drawLine(
                            (int)current_projection.x,
                            (int)current_projection.y,
                            (int)next_projection_x.x,
                            (int)next_projection_x.y,
                            1,
                            kColorBlack
                        );
                    }
                }
            }
            ++idx;
        }
    }

    for (int i = 0; i < WORLD_PARTICLE_COUNT; ++i) {
        ProjectedPoint particle_projection;
        if (!camera_project(camera, world->particles[i], &particle_projection)) {
            continue;
        }

        float radius = 200.0f / particle_projection.z;
        if (radius < 1.0f) {
            radius = 1.0f;
        }
        if (radius > 8.0f) {
            radius = 8.0f;
        }

        pd->graphics->fillEllipse(
            (int)(particle_projection.x - radius),
            (int)(particle_projection.y - radius),
            (int)(radius * 2.0f),
            (int)(radius * 2.0f),
            0.0f,
            0.0f,
            kColorBlack
        );
    }

    const float limit = (float)world->size - 5.0f;
    if (fabsf(camera->position.x) >= limit || fabsf(camera->position.y) >= limit || fabsf(camera->position.z) >= limit) {
        pd->graphics->drawText("WALL HIT", 8, kUTF8Encoding, 5, 220);
    }

}
